File: src/main/java/com/github/ccguyka/showupdates/filter/PluginFilter.java
package com.github.ccguyka.showupdates.filter;

import static java.util.stream.Collectors.toSet;

import java.io.IOException;
import java.util.Objects;
import java.util.Set;

import org.apache.maven.artifact.Artifact;
import org.apache.maven.plugin.logging.Log;
import org.apache.maven.project.MavenProject;

import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.github.ccguyka.showupdates.filter.pom.Project;

public class PluginFilter {

    private final MavenProject project;
    private final Log log;
    private final XmlMapper xmlMapper = new XmlMapper();

    public PluginFilter(MavenProject project, Log log) {
        this.project = project;
        this.log = log;
    }

    public Set<Artifact> filter(final Set<Artifact> artifacts) {
        try {
            Project pom = xmlMapper.readValue(project.getFile(), Project.class);
            return filter(artifacts, pom);
        } catch (final IOException e) {
            log.warn("Not able to read pom.xml file");
            return artifacts;
        }
    }

    private Set<Artifact> filter(Set<Artifact> artifacts, Project pom) {
        return artifacts.stream()
            .filter(dependency -> pom.getBuild().getPlugins()
                    .stream()
                    .anyMatch(dep -> Objects.equals(dep.getArtifactId(), dependency.getArtifactId())
                            && Objects.equals(dep.getGroupId(), dependency.getGroupId())))
            .collect(toSet());
}
}

File: src/main/kotlin/com/github/ccguyka/showupdates/filter/PluginFilter.kt
package com.github.ccguyka.showupdates.filter

import com.fasterxml.jackson.dataformat.xml.XmlMapper
import com.github.ccguyka.showupdates.filter.pom.Dependency
import com.github.ccguyka.showupdates.filter.pom.Project
import org.apache.maven.artifact.Artifact
import org.apache.maven.plugin.logging.Log
import org.apache.maven.project.MavenProject
import java.io.IOException
import java.util.stream.Collectors

class PluginFilter(private val project: MavenProject, private val log: Log) {

    private val xmlMapper = XmlMapper()

    fun filter(artifacts: Set<Artifact>): Set<Artifact> {
        return try {
            val pom = xmlMapper.readValue(project.file, Project::class.java)
            filter(artifacts, pom)
        } catch (e: IOException) {
            log.warn("Not able to read pom.xml file")
            artifacts
        }
    }

    private fun filter(artifacts: Set<Artifact>, pom: Project): Set<Artifact> {
        return artifacts.stream()
                .filter { dependency: Artifact ->
                    pom.build.plugins
                            .stream()
                            .anyMatch { dep: Dependency ->
                                (dep.artifactId == dependency.artifactId
                                        && dep.groupId == dependency.groupId)
                            }
                }
                .collect(Collectors.toSet())
    }
}
